src/EventStore.Core/Services/RedactionService.cs

```diff
@@ -0,0 +1,261 @@
+﻿using System;
+using System.IO;
+using System.Threading;
+using EventStore.Common.Utils;
+using EventStore.Core.Bus;
+using EventStore.Core.Data.Redaction;
+using EventStore.Core.Exceptions;
+using EventStore.Core.Messages;
+using EventStore.Core.Messaging;
+using EventStore.Core.Services.Storage.ReaderIndex;
+using EventStore.Core.Synchronization;
+using EventStore.Core.TransactionLog.Chunks;
+using EventStore.Core.TransactionLog.Chunks.TFChunk;
+using Serilog;
+
+namespace EventStore.Core.Services {
+	public class RedactionService<TStreamId> :
+		IHandle<RedactionMessage.GetEventPosition>,
+		IHandle<RedactionMessage.SwitchChunkLock>,
+		IHandle<RedactionMessage.SwitchChunk>,
+		IHandle<RedactionMessage.SwitchChunkUnlock>,
+		IHandle<SystemMessage.BecomeShuttingDown> {
+
+		private readonly IQueuedHandler _queuedHandler;
+		private readonly TFChunkDb _db;
+		private readonly IReadIndex<TStreamId> _readIndex;
+		private readonly SemaphoreSlimLock _switchChunksLock;
+
+		private const string NewChunkFileExtension = ".tmp";
+
+		public RedactionService(
+			IQueuedHandler queuedHandler,
+			TFChunkDb db,
+			IReadIndex<TStreamId> readIndex,
+			SemaphoreSlimLock switchChunksLock) {
+			Ensure.NotNull(queuedHandler, nameof(queuedHandler));
+			Ensure.NotNull(db, nameof(db));
+			Ensure.NotNull(readIndex, nameof(readIndex));
+			Ensure.NotNull(switchChunksLock, nameof(switchChunksLock));
+
+			_queuedHandler = queuedHandler;
```

Comment:
perhaps don't need to pass this in here solely to stop it on shutdown. since the bootstrapper is doing the subscriptions for us it could probably also take care of shutting down the queue too?

Reply:
yup, i agree, it's not too neat in this case. i've actually followed the pattern used by other services: e.g SubscriptionsService and PersistentSubscriptionService. it seems that the advantage of doing it this way is that you can stop other internal things before stopping the queue (e.g as done in PersistentSubscriptionService)

Reply:
aha ok :+1: